.gitlab-ci.yml

```diff
@@ -42,7 +42,7 @@ pull-request-validation:
     - source $CI_PROJECT_DIR/share/ci/git_merge.sh
     - pip3 install -U flake8 pyflakes
```

Comment:
Does it make sense to remove the explicit tests here and use `git-pre-commit`. This should prevent our tests and the rules in pre-commit will be de-sync at some point.

Reply:
We can only use pre-commit if we can test only without changing the code.

Reply:
I think the feature will not come https://github.com/pre-commit/pre-commit/issues/2144 but we can run `pre-commit` and then check if we have unstaged files.

Reply:
But the discussion suggests it's pretty straightforward as we are already doing short-circuiting:
1. Install `pre-commit`
2. Run `pre-commit`
3. If `EXIT_CODE != 0`: fail pipeline, else we can be sure that running `pre-commit` didn't actually change anything and we can continue as usual.

(There could theoretically be some really badly designed hooks that apply changes without reporting a failure to `pre-commit` but that sounds like sufficiently bad design to reject them on these grounds alone.)

Reply:
Following my other comment, I think we should move this discussion into #4787.